packages/medusa/src/api/middlewares/transform-query.ts

```diff
@@ -77,3 +69,67 @@ export function transformQuery<
     }
   }
 }
+
+/**
+ * Build the store allowed props based on the custom fields query params, the defaults and the includes options.
+ * This can be used later with `cleanResponseData` in order to clean up the returned objects to the client.
+ * @param queryConfig
+ * @param validated
+ * @param includesOptions
+ */
+function getStoreAllowedProperties<TEntity extends BaseEntity>(
+  validated: RequestQueryFields,
+  includesOptions: Record<string, boolean>,
+  queryConfig?: QueryConfig<TEntity>
+): string[] {
+  const allowed: string[] = []
+  if (
+    (queryConfig?.defaultFields || validated.fields) &&
+    (queryConfig?.defaultRelations || validated.expand)
+  ) {
+    allowed.push(
+      ...(validated.fields
+        ? validated.fields.split(",")
+        : queryConfig?.allowedFields || []),
+      ...(validated.expand
+        ? validated.expand.split(",")
+        : queryConfig?.allowedRelations || [])
+    )
+  }
+
+  const includeKeys = Object.keys(includesOptions)
+  if (includeKeys) {
+    allowed.push(...includeKeys)
+  }
+
+  return allowed
+}
+
+/**
+ * Build the admin allowed props based on the custom fields query params, the defaults and the includes options.
+ * Since admin can access everything, it is only in order to return what the user asked for through fields and expand query params.
+ * This can be used later with `cleanResponseData` in order to clean up the returned objects to the client.
+ * @param queryConfig
+ * @param validated
+ * @param includesOptions
+ */
+function getAdminAllowedProperties<TEntity extends BaseEntity>(
```

Comment:
thought: I feel the allowed/excluded logic should lie in the routes file and then, as a part of the clean response function, it should accept an excluded or allowed array that will do the exclusion. wdyt?

Reply:
I just refactored the middleware by creating local utility methods